zaza/utilities/__init__.py

```diff
@@ -13,3 +13,113 @@
+import json
+import os
+import types
+
+
+class ConfigurableMixin:
+    """A helper class that can configure members.
+
+    If a class inherits from this class as a mixin, then it gets a 'configure'
+    function that can configure members of the object.
+
+    e.g.
+
+        class SomeThing(ConfigurableMixin):
+
+            def __init__(self, *args, **kwargs):
+                self.a = None
+                self.b = None
+                self.configure(**kwargs)
+
+
+        x = Something(a=4)
+        x.a => 4
+        x.configure(b=9)
+        x.b => 9
+
+    """
+
+    def configure(self, **kwargs):
+        """Configure the non-private attributes of the object.
+
+        The keyword args can be used to configure the attributes of the object
+        directly.
+        """
+        for k, v in kwargs.items():
+            if k.startswith("_"):
+                raise AttributeError("Key {} is private in {}"
+                                     .format(k, self.__class__.__name__))
+            try:
+                _type = getattr(self.__class__, k, None)
+                if isinstance(_type, (property, types.FunctionType)):
+                    raise AttributeError()
+                getattr(self, k)
+                setattr(self, k, v)
+            except AttributeError:
+                raise AttributeError(
+                    "key {} doesn't belong to {} or is a method/property"
+                    .format(k, self.__class__.__name__))
+
+
+cache = {}
+
+
+def cached(func):
+    """Cache return values for multiple executions of func + args.
+
+    For example::
+
+        @cached
+        def unit_get(attribute):
+            pass
+
+        unit_get('test')
+
+    will cache the result of unit_get + 'test' for future calls.
+
+    :param func: the function to cache
+    :type func: Callable[_T]
+    :returns: Callable[_T]
+    """
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        global cache
+        key = json.dumps((func, args, kwargs), sort_keys=True, default=str)
+        try:
+            return cache[key]
+        except KeyError:
+            pass  # Drop out of the exception handler scope.
+        res = func(*args, **kwargs)
+        cache[key] = res
+        return res
+    wrapper._wrapped = func
+    return wrapper
+
+
+def expand_vars(context, value):
+    """Search the variable and see if variables need to be expanded.
+
+    This expands ${ENV} and {context} variables in a :param:`value` parameter.
+
+    # :param env_deployment: the deployment parameters.
+    # :type env_deployment: environmentdeploy
+    :param context: a context of dictionary keys for filling in values
+    :type context: Dict[str, str]
+    :param value: the value to do variable expansion on.
+    :type value: str
+    :returns: the expanded string
+    :rtype: str
+    """
+    if not isinstance(value, str):
+        return value
+    value = os.path.expandvars(value)
+    for k, v in context.items():
+        var = "{" + k + "}"
```

Comment:
This looks a bit white space sensitive. I believe f strings will resolve {k} and { k } and people might expect a similar behaviour

Reply:
I see your issue with being whitespace sensitive, but I believe I can't use f-strings as we still have to rung zaza on xenial runners for the moment, AFAICT.  I'll re-work this bit of code to be less whitespace sensitive.